pkg/kclient/all.go

```diff
@@ -0,0 +1,182 @@
+			continue
+		}
+		wg.Add(1)
+		go func(a apiResource) {
+			defer wg.Done()
+			klog.V(4).Infof("[query api] start: %s", a.GroupVersionResource())
+			v, err := queryAPI(client, a, ns, selector)
+			if err != nil {
+				klog.V(4).Infof("[query api] error querying: %s, error=%v", a.GroupVersionResource(), err)
+				errResult = err
+				return
+			}
+			mu.Lock()
+			out = append(out, v...)
+			mu.Unlock()
+			klog.V(4).Infof("[query api]  done: %s, found %d apis", a.GroupVersionResource(), len(v))
+		}(api)
+	}
+
+	klog.V(2).Infof("fired up all goroutines to query APIs")
+	wg.Wait()
+	klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
+	klog.V(2).Infof("query result: error=%v, objects=%d", errResult, len(out))
+	return out, errResult
+}
+
+func queryAPI(client dynamic.Interface, api apiResource, ns string, selector string) ([]unstructured.Unstructured, error) {
+	var out []unstructured.Unstructured
+
+	var next string
+	for {
+		var intf dynamic.ResourceInterface
+		nintf := client.Resource(api.GroupVersionResource())
+		intf = nintf.Namespace(ns)
+		resp, err := intf.List(context.TODO(), metav1.ListOptions{
+			Limit:         250,
+			Continue:      next,
+			LabelSelector: selector,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("listing resources failed (%s): %w", api.GroupVersionResource(), err)
+		}
+		out = append(out, resp.Items...)
+
+		next = resp.GetContinue()
+		if next == "" {
+			break
+		}
+	}
+	return out, nil
+}
+
+type apiResource struct {
+	r  metav1.APIResource
+	gv schema.GroupVersion
+}
+
+func (a apiResource) GroupVersionResource() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    a.gv.Group,
+		Version:  a.gv.Version,
+		Resource: a.r.Name,
+	}
+}
+
+type resourceNameLookup map[string][]apiResource
+
+type resourceMap struct {
+	list []apiResource
+	m    resourceNameLookup
+}
+
+func findAPIs(client discovery.DiscoveryInterface) (*resourceMap, error) {
+	start := time.Now()
+	resList, err := client.ServerPreferredResources()
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch api groups from kubernetes: %w", err)
+	}
+	klog.V(2).Infof("queried api discovery in %v", time.Since(start))
+	klog.V(3).Infof("found %d items (groups) in server-preferred APIResourceList", len(resList))
+
+	rm := &resourceMap{
+		m: make(resourceNameLookup),
+	}
+	for _, group := range resList {
+		klog.V(5).Infof("iterating over group %s/%s (%d apis)", group.GroupVersion, group.APIVersion, len(group.APIResources))
+		gv, err := schema.ParseGroupVersion(group.GroupVersion)
+		if err != nil {
+			return nil, fmt.Errorf("%q cannot be parsed into groupversion: %w", group.GroupVersion, err)
+		}
+
+		for _, apiRes := range group.APIResources {
+			klog.V(5).Infof("  api=%s namespaced=%v", apiRes.Name, apiRes.Namespaced)
+			if !contains(apiRes.Verbs, "list") {
+				klog.V(4).Infof("    api (%s) doesn't have required verb, skipping: %v", apiRes.Name, apiRes.Verbs)
```

Comment:
Being pedantic/nitpicky here. Apologies in advance.

> `doesn't have required verb, skipping`

If something is "required" but it doesn't exist, we should fail. In this case, "list" doesn't fall into the definition of required, IMO.